pkg/virtctl/imageupload/imageupload.go

```diff
@@ -291,12 +306,36 @@ func ConstructUploadProxyPath(uploadProxyURL string) (string, error) {
 	if !strings.Contains(uploadProxyURL, UploadProxyURI) {
 		u.Path = path.Join(u.Path, UploadProxyURI)
 	}
+	return u.String(), nil
+}
+
+//ConstructUploadProxyPathAsync - receives uploadproxy adress and concatenates to it URI
+func ConstructUploadProxyPathAsync(uploadProxyURL, token string, insecure bool) (string, error) {
+	u, err := url.Parse(uploadProxyURL)
+
+	if err != nil {
+		return "", err
+	}
+
+	if !strings.Contains(uploadProxyURL, UploadProxyURIAsync) {
+		u.Path = path.Join(u.Path, UploadProxyURIAsync)
+	}
+
+	// Attempt to discover async URL
+	client := httpClientCreatorFunc(insecure)
+	req, _ := http.NewRequest("HEAD", u.String(), nil)
+	req.Header.Add("Authorization", "Bearer "+token)
+	resp, err := client.Do(req)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		// Async not available, use regular upload url.
```

Comment:
No if the token is incorrect, it does mean the end point is unavailable since the endpoint is protected by the token. The code will then try the non async endpoint and fail there as well and return a token error as before.

Reply:
Ack, thanks for explaining :smiley: